File: plugins/lib-connect/Protocol/BinaryMessage.cs
using System;
using System.Text;

namespace BeyondImmersion.BannouService.Connect.Protocol;

/// <summary>
/// Represents a complete binary message in the Bannou WebSocket protocol.
/// Request messages use a 31-byte header, response messages use a 16-byte header.
/// Provides zero-copy parsing and serialization for optimal performance.
/// </summary>
public readonly struct BinaryMessage
{
    /// <summary>
    /// Size of the binary header in bytes for request messages.
    /// </summary>
    public const int HeaderSize = 31;

    /// <summary>
    /// Size of the binary header in bytes for response messages.
    /// Response headers are smaller because ServiceGuid is not needed (client correlates via MessageId).
    /// </summary>
    public const int ResponseHeaderSize = 16;

    /// <summary>
    /// Message behavior flags
    /// </summary>
    public MessageFlags Flags { get; }

    /// <summary>
    /// Channel for sequential message processing (0-65535, 0 = default)
    /// </summary>
    public ushort Channel { get; }

    /// <summary>
    /// Per-channel sequence number for message ordering
    /// </summary>
    public uint SequenceNumber { get; }

    /// <summary>
    /// Client-salted service GUID for routing (only used in request messages).
    /// For response messages, this is Guid.Empty since routing is not needed.
    /// </summary>
    public Guid ServiceGuid { get; }

    /// <summary>
    /// Unique message ID for request/response correlation
    /// </summary>
    public ulong MessageId { get; }

    /// <summary>
    /// Response code for response messages (0 = OK, non-zero = error).
    /// Only meaningful when Flags has Response set.
    /// </summary>
    public byte ResponseCode { get; }

    /// <summary>
    /// Message payload (JSON or binary data based on flags).
    /// For error responses (non-zero ResponseCode), this is empty.
    /// </summary>
    public ReadOnlyMemory<byte> Payload { get; }

    /// <summary>
    /// Initializes a new binary message with the specified parameters (for requests).
    /// </summary>
    public BinaryMessage(
        MessageFlags flags,
        ushort channel,
        uint sequenceNumber,
        Guid serviceGuid,
        ulong messageId,
        ReadOnlyMemory<byte> payload)
    {
        Flags = flags;
        Channel = channel;
        SequenceNumber = sequenceNumber;
        ServiceGuid = serviceGuid;
        MessageId = messageId;
        ResponseCode = 0;
        Payload = payload;
    }

    /// <summary>
    /// Initializes a new binary message with the specified parameters (for responses).
    /// </summary>
    public BinaryMessage(
        MessageFlags flags,
        ushort channel,
        uint sequenceNumber,
        ulong messageId,
        byte responseCode,
        ReadOnlyMemory<byte> payload)
    {
        Flags = flags;
        Channel = channel;
        SequenceNumber = sequenceNumber;
        ServiceGuid = Guid.Empty; // Not used in responses
        MessageId = messageId;
        ResponseCode = responseCode;
        Payload = payload;
    }

    /// <summary>
    /// Creates a binary message from a JSON string payload.
    /// </summary>
    public static BinaryMessage FromJson(
        ushort channel,
        uint sequenceNumber,
        Guid serviceGuid,
        ulong messageId,
        string jsonPayload,
        MessageFlags additionalFlags = MessageFlags.None)
    {
        var payloadBytes = Encoding.UTF8.GetBytes(jsonPayload);
        return new BinaryMessage(
            additionalFlags, // JSON is default, no Binary flag needed
            channel,
            sequenceNumber,
            serviceGuid,
            messageId,
            payloadBytes);
    }

    /// <summary>
    /// Creates a binary message from raw binary payload.
    /// </summary>
    public static BinaryMessage FromBinary(
        ushort channel,
        uint sequenceNumber,
        Guid serviceGuid,
        ulong messageId,
        ReadOnlyMemory<byte> binaryPayload,
        MessageFlags additionalFlags = MessageFlags.None)
    {
        return new BinaryMessage(
            additionalFlags | MessageFlags.Binary, // Mark as binary
            channel,
            sequenceNumber,
            serviceGuid,
            messageId,
            binaryPayload);
    }

    /// <summary>
    /// Creates a response message for the given request.
    /// For error responses (non-zero responseCode), the payload should be empty.
    /// For success responses (responseCode = 0), the payload contains the response data.
    /// </summary>
    public static BinaryMessage CreateResponse(
        BinaryMessage request,
        ResponseCodes responseCode,
        ReadOnlyMemory<byte> responsePayload = default)
    {
        var flags = request.Flags | MessageFlags.Response;

        // For error responses, payload is empty - the response code tells the story
        // For success responses, payload contains the actual response data
        var payload = responseCode != ResponseCodes.OK ? ReadOnlyMemory<byte>.Empty : responsePayload;

        return new BinaryMessage(
            flags,
            request.Channel,
            request.SequenceNumber,
            request.MessageId,
            (byte)responseCode,
            payload);
    }

    /// <summary>
    /// Gets the JSON payload as a string (if the message is JSON).
    /// </summary>
    public string GetJsonPayload()
    {
        if (Flags.HasFlag(MessageFlags.Binary))
        {
            throw new InvalidOperationException("Cannot get JSON payload from binary message");
        }

        return Encoding.UTF8.GetString(Payload.Span);
    }

    /// <summary>
    /// Parses a binary message from a byte array.
    /// Automatically detects whether this is a request (31-byte header) or response (16-byte header).
    /// </summary>
    /// <param name="buffer">Buffer containing the complete message</param>
    /// <param name="messageLength">Total length of the message</param>
    /// <returns>Parsed binary message</returns>
    public static BinaryMessage Parse(byte[] buffer, int messageLength)
    {
        if (messageLength < 1)
        {
            throw new ArgumentException("Message too short. Expected at least 1 byte for flags.");
        }

        var flags = (MessageFlags)buffer[0];

        // Check if this is a response message (16-byte header) or request message (31-byte header)
        if (flags.HasFlag(MessageFlags.Response))
        {
            return ParseResponse(buffer, messageLength, flags);
        }
        else
        {
            return ParseRequest(buffer, messageLength, flags);
        }
    }

    private static BinaryMessage ParseRequest(byte[] buffer, int messageLength, MessageFlags flags)
    {
        if (messageLength < HeaderSize)
        {
            throw new ArgumentException($"Request message too short. Expected at least {HeaderSize} bytes, got {messageLength}");
        }

        // Parse 31-byte request header using consistent network byte order
        var channel = NetworkByteOrder.ReadUInt16(buffer.AsSpan(1, 2));        // Bytes 1-2: Channel
        var sequence = NetworkByteOrder.ReadUInt32(buffer.AsSpan(3, 4));       // Bytes 3-6: Sequence number
        var serviceGuid = NetworkByteOrder.ReadGuid(buffer.AsSpan(7, 16));     // Bytes 7-22: Service GUID
        var messageId = NetworkByteOrder.ReadUInt64(buffer.AsSpan(23, 8));     // Bytes 23-30: Message ID

        // Extract payload (remaining bytes after 31-byte header)
        var payloadLength = messageLength - HeaderSize;
        var payload = new byte[payloadLength];
        if (payloadLength > 0)
        {
            Array.Copy(buffer, HeaderSize, payload, 0, payloadLength);
        }

        return new BinaryMessage(flags, channel, sequence, serviceGuid, messageId, payload);
    }

    private static BinaryMessage ParseResponse(byte[] buffer, int messageLength, MessageFlags flags)
    {
        if (messageLength < ResponseHeaderSize)
        {
            throw new ArgumentException($"Response message too short. Expected at least {ResponseHeaderSize} bytes, got {messageLength}");
        }

        // Parse 16-byte response header using consistent network byte order
        var channel = NetworkByteOrder.ReadUInt16(buffer.AsSpan(1, 2));        // Bytes 1-2: Channel
        var sequence = NetworkByteOrder.ReadUInt32(buffer.AsSpan(3, 4));       // Bytes 3-6: Sequence number
        var messageId = NetworkByteOrder.ReadUInt64(buffer.AsSpan(7, 8));      // Bytes 7-14: Message ID
        var responseCode = buffer[15];                                          // Byte 15: Response code

        // Extract payload (remaining bytes after 16-byte header)
        var payloadLength = messageLength - ResponseHeaderSize;
        var payload = new byte[payloadLength];
        if (payloadLength > 0)
        {
            Array.Copy(buffer, ResponseHeaderSize, payload, 0, payloadLength);
        }

        return new BinaryMessage(flags, channel, sequence, messageId, responseCode, payload);
    }

    /// <summary>
    /// Serializes the binary message to a byte array.
    /// Uses 31-byte header for requests, 16-byte header for responses.
    /// </summary>
    /// <returns>Complete message as byte array</returns>
    public byte[] ToByteArray()
    {
        if (IsResponse)
        {
            return ToResponseByteArray();
        }
        else
        {
            return ToRequestByteArray();
        }
    }

    private byte[] ToRequestByteArray()
    {
        var totalLength = HeaderSize + Payload.Length;
        var result = new byte[totalLength];

        // Build 31-byte request header using consistent network byte order
        result[0] = (byte)Flags;                                              // Byte 0: Message flags
        NetworkByteOrder.WriteUInt16(result.AsSpan(1, 2), Channel);          // Bytes 1-2: Channel
        NetworkByteOrder.WriteUInt32(result.AsSpan(3, 4), SequenceNumber);   // Bytes 3-6: Sequence
        NetworkByteOrder.WriteGuid(result.AsSpan(7, 16), ServiceGuid);       // Bytes 7-22: Service GUID
        NetworkByteOrder.WriteUInt64(result.AsSpan(23, 8), MessageId);       // Bytes 23-30: Message ID

        // Append payload
        if (!Payload.IsEmpty)
        {
            Payload.Span.CopyTo(result.AsSpan(HeaderSize));
        }

        return result;
    }

    private byte[] ToResponseByteArray()
    {
        var totalLength = ResponseHeaderSize + Payload.Length;
        var result = new byte[totalLength];

        // Build 16-byte response header using consistent network byte order
        result[0] = (byte)Flags;                                              // Byte 0: Message flags
        NetworkByteOrder.WriteUInt16(result.AsSpan(1, 2), Channel);          // Bytes 1-2: Channel
        NetworkByteOrder.WriteUInt32(result.AsSpan(3, 4), SequenceNumber);   // Bytes 3-6: Sequence
        NetworkByteOrder.WriteUInt64(result.AsSpan(7, 8), MessageId);        // Bytes 7-14: Message ID
        result[15] = ResponseCode;                                            // Byte 15: Response code

        // Append payload (empty for error responses)
        if (!Payload.IsEmpty)
        {
            Payload.Span.CopyTo(result.AsSpan(ResponseHeaderSize));
        }

        return result;
    }

    /// <summary>
    /// Returns true if this message expects a response.
    /// </summary>
    public bool ExpectsResponse => !Flags.HasFlag(MessageFlags.Event) && !Flags.HasFlag(MessageFlags.Response);

    /// <summary>
    /// Returns true if this message is a response to another message.
    /// </summary>
    public bool IsResponse => Flags.HasFlag(MessageFlags.Response);

    /// <summary>
    /// Returns true if this message should be routed to a client (not a service).
    /// </summary>
    public bool IsClientRouted => Flags.HasFlag(MessageFlags.Client);

    /// <summary>
    /// Returns true if this is a successful response (ResponseCode is 0).
    /// Only meaningful for response messages.
    /// </summary>
    public bool IsSuccess => IsResponse && ResponseCode == 0;

    /// <summary>
    /// Returns true if this is an error response (ResponseCode is non-zero).
    /// Only meaningful for response messages.
    /// </summary>
    public bool IsError => IsResponse && ResponseCode != 0;

    /// <summary>
    /// Returns true if this message requests endpoint metadata instead of executing the endpoint.
    /// When set, the Channel field specifies the meta type:
    /// 0 = endpoint-info, 1 = request-schema, 2 = response-schema, 3 = full-schema.
    /// </summary>
    public bool IsMeta => Flags.HasFlag(MessageFlags.Meta);

    /// <summary>
    /// Returns a string representation of the message for debugging.
    /// </summary>
    public override string ToString()
    {
        if (IsResponse)
        {
            return $"BinaryMessage(Response, Flags={Flags}, Channel={Channel}, Seq={SequenceNumber}, " +
                    $"MessageId={MessageId}, ResponseCode={ResponseCode}, PayloadSize={Payload.Length})";
        }
        else
        {
            return $"BinaryMessage(Request, Flags={Flags}, Channel={Channel}, Seq={SequenceNumber}, " +
                    $"ServiceGuid={ServiceGuid}, MessageId={MessageId}, PayloadSize={Payload.Length})";
        }
    }
}

File: sdks/typescript/client/src/protocol/BinaryMessage.ts
/**
 * Represents a complete binary message in the Bannou WebSocket protocol.
 * Request messages use a 31-byte header, response messages use a 16-byte header.
 * Provides zero-copy parsing and serialization for optimal performance.
 */

import { MessageFlags, hasFlag, isResponse as checkIsResponse } from './MessageFlags.js';
import {
  readUInt16,
  readUInt32,
  readUInt64,
  readGuid,
  writeUInt16,
  writeUInt32,
  writeUInt64,
  writeGuid,
  EMPTY_GUID,
} from './NetworkByteOrder.js';

/**
 * Size of the binary header in bytes for request messages.
 *
 * Request header layout (31 bytes):
 * - [0]     Flags (1 byte)
 * - [1-2]   Channel (uint16 BE)
 * - [3-6]   Sequence (uint32 BE)
 * - [7-22]  ServiceGUID (16 bytes, RFC 4122)
 * - [23-30] MessageID (uint64 BE)
 * - [31+]   Payload (JSON UTF-8)
 */
export const HEADER_SIZE = 31;

/**
 * Size of the binary header in bytes for response messages.
 * Response headers are smaller because ServiceGuid is not needed (client correlates via MessageId).
 *
 * Response header layout (16 bytes):
 * - [0]     Flags (1 byte, Response flag set)
 * - [1-2]   Channel (uint16 BE)
 * - [3-6]   Sequence (uint32 BE)
 * - [7-14]  MessageID (uint64 BE)
 * - [15]    ResponseCode (1 byte)
 * - [16+]   Payload (empty for errors)
 */
export const RESPONSE_HEADER_SIZE = 16;

/**
 * Represents a parsed or constructed binary message.
 */
export interface BinaryMessage {
  /** Message behavior flags */
  flags: number;

  /** Channel for sequential message processing (0-65535, 0 = default) */
  channel: number;

  /** Per-channel sequence number for message ordering */
  sequenceNumber: number;

  /** Client-salted service GUID for routing (only used in request messages) */
  serviceGuid: string;

  /** Unique message ID for request/response correlation */
  messageId: bigint;

  /** Response code for response messages (0 = OK, non-zero = error) */
  responseCode: number;

  /** Message payload (JSON or binary data based on flags) */
  payload: Uint8Array;
}

/**
 * Creates a request message.
 */
export function createRequest(
  flags: number,
  channel: number,
  sequenceNumber: number,
  serviceGuid: string,
  messageId: bigint,
  payload: Uint8Array
): BinaryMessage {
  return {
    flags,
    channel,
    sequenceNumber,
    serviceGuid,
    messageId,
    responseCode: 0,
    payload,
  };
}

/**
 * Creates a response message.
 */
export function createResponse(
  flags: number,
  channel: number,
  sequenceNumber: number,
  messageId: bigint,
  responseCode: number,
  payload: Uint8Array
): BinaryMessage {
  return {
    flags: flags | MessageFlags.Response,
    channel,
    sequenceNumber,
    serviceGuid: EMPTY_GUID,
    messageId,
    responseCode,
    payload,
  };
}

/**
 * Creates a binary message from a JSON string payload.
 */
export function fromJson(
  channel: number,
  sequenceNumber: number,
  serviceGuid: string,
  messageId: bigint,
  jsonPayload: string,
  additionalFlags: number = MessageFlags.None
): BinaryMessage {
  const payloadBytes = new TextEncoder().encode(jsonPayload);
  return createRequest(
    additionalFlags,
    channel,
    sequenceNumber,
    serviceGuid,
    messageId,
    payloadBytes
  );
}

/**
 * Gets the JSON payload as a string (if the message is JSON).
 * @throws Error if the message has the Binary flag set
 */
export function getJsonPayload(message: BinaryMessage): string {
  if (hasFlag(message.flags, MessageFlags.Binary)) {
    throw new Error('Cannot get JSON payload from binary message');
  }
  return new TextDecoder().decode(message.payload);
}

/**
 * Parses a binary message from a byte array.
 * Automatically detects whether this is a request (31-byte header) or response (16-byte header).
 */
export function parse(buffer: ArrayBuffer | Uint8Array): BinaryMessage {
  const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);

  if (bytes.length < 1) {
    throw new Error('Message too short. Expected at least 1 byte for flags.');
  }

  const flags = view.getUint8(0);

  // Check if this is a response message (16-byte header) or request message (31-byte header)
  if (checkIsResponse(flags)) {
    return parseResponse(view, bytes);
  } else {
    return parseRequest(view, bytes);
  }
}

function parseRequest(view: DataView, bytes: Uint8Array): BinaryMessage {
  if (bytes.length < HEADER_SIZE) {
    throw new Error(
      `Request message too short. Expected at least ${HEADER_SIZE} bytes, got ${bytes.length}`
    );
  }

  const flags = view.getUint8(0);
  const channel = readUInt16(view, 1);
  const sequenceNumber = readUInt32(view, 3);
  const serviceGuid = readGuid(view, 7);
  const messageId = readUInt64(view, 23);

  // Extract payload (remaining bytes after 31-byte header)
  const payload = bytes.slice(HEADER_SIZE);

  return {
    flags,
    channel,
    sequenceNumber,
    serviceGuid,
    messageId,
    responseCode: 0,
    payload,
  };
}

function parseResponse(view: DataView, bytes: Uint8Array): BinaryMessage {
  if (bytes.length < RESPONSE_HEADER_SIZE) {
    throw new Error(
      `Response message too short. Expected at least ` +
        `${RESPONSE_HEADER_SIZE} bytes, got ${bytes.length}`
    );
  }

  const flags = view.getUint8(0);
  const channel = readUInt16(view, 1);
  const sequenceNumber = readUInt32(view, 3);
  const messageId = readUInt64(view, 7);
  const responseCode = view.getUint8(15);

  // Extract payload (remaining bytes after 16-byte header)
  const payload = bytes.slice(RESPONSE_HEADER_SIZE);

  return {
    flags,
    channel,
    sequenceNumber,
    serviceGuid: EMPTY_GUID,
    messageId,
    responseCode,
    payload,
  };
}

/**
 * Serializes the binary message to a byte array.
 * Uses 31-byte header for requests, 16-byte header for responses.
 */
export function toByteArray(message: BinaryMessage): Uint8Array {
  if (checkIsResponse(message.flags)) {
    return toResponseByteArray(message);
  } else {
    return toRequestByteArray(message);
  }
}

function toRequestByteArray(message: BinaryMessage): Uint8Array {
  const totalLength = HEADER_SIZE + message.payload.length;
  const result = new Uint8Array(totalLength);
  const view = new DataView(result.buffer);

  // Build 31-byte request header
  view.setUint8(0, message.flags);
  writeUInt16(view, 1, message.channel);
  writeUInt32(view, 3, message.sequenceNumber);
  writeGuid(view, 7, message.serviceGuid);
  writeUInt64(view, 23, message.messageId);

  // Append payload
  if (message.payload.length > 0) {
    result.set(message.payload, HEADER_SIZE);
  }

  return result;
}

function toResponseByteArray(message: BinaryMessage): Uint8Array {
  const totalLength = RESPONSE_HEADER_SIZE + message.payload.length;
  const result = new Uint8Array(totalLength);
  const view = new DataView(result.buffer);

  // Build 16-byte response header
  view.setUint8(0, message.flags);
  writeUInt16(view, 1, message.channel);
  writeUInt32(view, 3, message.sequenceNumber);
  writeUInt64(view, 7, message.messageId);
  view.setUint8(15, message.responseCode);

  // Append payload (empty for error responses)
  if (message.payload.length > 0) {
    result.set(message.payload, RESPONSE_HEADER_SIZE);
  }

  return result;
}

/**
 * Returns true if this message expects a response.
 */
export function expectsResponse(message: BinaryMessage): boolean {
  return !hasFlag(message.flags, MessageFlags.Event) && !checkIsResponse(message.flags);
}

/**
 * Returns true if this message is a response to another message.
 */
export function isResponse(message: BinaryMessage): boolean {
  return checkIsResponse(message.flags);
}

/**
 * Returns true if this message should be routed to a client (not a service).
 */
export function isClientRouted(message: BinaryMessage): boolean {
  return hasFlag(message.flags, MessageFlags.Client);
}

/**
 * Returns true if this is a successful response (ResponseCode is 0).
 */
export function isSuccess(message: BinaryMessage): boolean {
  return checkIsResponse(message.flags) && message.responseCode === 0;
}

/**
 * Returns true if this is an error response (ResponseCode is non-zero).
 */
export function isError(message: BinaryMessage): boolean {
  return checkIsResponse(message.flags) && message.responseCode !== 0;
}

/**
 * Returns true if this message requests endpoint metadata.
 */
export function isMeta(message: BinaryMessage): boolean {
  return hasFlag(message.flags, MessageFlags.Meta);
}
